File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/payment/infrastructure/PaymentCardAdapterImpl.java
package com.example.accountservice.payment.infrastructure;

import com.example.accountservice.card.application.query.CardQuery;
import com.example.accountservice.card.domain.Card;
import com.example.accountservice.card.domain.CardFindQuery;
import com.example.accountservice.card.domain.CardStatus;
import com.example.accountservice.payment.application.adapter.CardAdapter;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

/**
 * The implementation (ACL) of {@link CardAdapter}. Injects and calls the read interface ({@link
 * CardQuery}) exposed by the Card BC, and translates the Card BC's model into the minimal shape the
 * Payment BC uses ({@link CardAdapter.CardView}). This is the sole boundary point that keeps Card's
 * Repository/domain objects from being exposed to Payment's Application/Domain layers (the same
 * structure as card/infrastructure/AccountAdapterImpl.java).
 */
// The reason it's named PaymentCardAdapterImpl is the same as PaymentAccountAdapterImpl — this
// leaves no room for another BC to later add a different implementation under the same simple
// class name (CardAdapterImpl).
@Component
@RequiredArgsConstructor
public class PaymentCardAdapterImpl implements CardAdapter {

    private final CardQuery cardQuery;

    @Override
    public Optional<CardView> findCard(String cardId, String ownerId) {
        return cardQuery
                .findCards(new CardFindQuery(0, 1, cardId, ownerId, null, null))
                .cards()
                .stream()
                .findFirst()
                .map(this::toCardView);
    }

    private CardView toCardView(Card card) {
        return new CardView(
                card.getCardId(), card.getAccountId(), card.getStatus() == CardStatus.ACTIVE);
    }
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/payment/infrastructure/PaymentCardAdapterImpl.kt
package com.example.accountservice.payment.infrastructure

import com.example.accountservice.card.application.query.CardQuery
import com.example.accountservice.card.domain.CardFindQuery
import com.example.accountservice.card.domain.CardStatus
import com.example.accountservice.payment.application.adapter.CardAdapter
import com.example.accountservice.payment.application.adapter.CardView
import org.springframework.stereotype.Component

/**
 * The implementation of [CardAdapter] (ACL). Injects and calls the read port ([CardQuery]) exposed by
 * Card BC, and translates Card BC's model ([com.example.accountservice.card.domain.Card]·[CardStatus])
 * into the minimal shape Payment BC uses ([CardView]). It does not reference Card's write
 * Repository/domain methods.
 *
 * Card's "no card" signal is `CardQuery.findCards` returning an empty list (a single-record lookup is
 * also handled via `take = 1` + `firstOrNull()`), and this is propagated as-is into the `null` that the
 * Payment domain understands — Card's exception types (CardNotFoundException, etc.) do not leak into
 * the Payment layer.
 *
 * See the [com.example.accountservice.payment.infrastructure.PaymentAccountAdapterImpl] doc comment
 * for why the class name carries the `Payment` prefix — there is currently no other `CardAdapterImpl`
 * that would collide with this name, but the same prefix convention is applied consistently anyway.
 */
@Component
class PaymentCardAdapterImpl(
    private val cardQuery: CardQuery,
) : CardAdapter {
    override fun findCard(
        cardId: String,
        ownerId: String,
    ): CardView? {
        val (cards, _) = cardQuery.findCards(CardFindQuery(page = 0, take = 1, cardId = cardId, ownerId = ownerId))
        return cards.firstOrNull()?.let { card ->
            CardView(cardId = card.cardId, accountId = card.accountId, active = card.status == CardStatus.ACTIVE)
        }
    }
}
